components/cam/src/dynamics/se/dyn_grid.F90

```diff
@@ -1174,7 +1538,116 @@ subroutine dyn_grid_get_elem_coords( ie, rlon, rlat, cdex )
+    ! matter here, but it should generally always be counter-clockwise.
+    corner_offset_i = (/-1.,1.,1.,-1./)
+    corner_offset_j = (/-1.,-1.,1.,1./)
+    
+    ! calculate coordinates and area for local blocks
+    do ie = 1, nelemd
+      do j = 1, fv_nphys
+        do i = 1, fv_nphys
+          !-------------------------------------------------------------------
+          ! FV coordinates are defined on the reference element defined 
+          ! over [-1,1] and then mapped to the sphere using ref2sphere
+          ! The code below is a reduction of a general formula to define
+          ! midpoints m of n cells over a range [a,b]:
+          ! m(i) = a + (b-a)/n * ( i - 0.5 )      where i = 1,n
+          !-------------------------------------------------------------------
+          ref_i = -1._lng_dbl + 2._lng_dbl/real(fv_nphys,lng_dbl) & 
+                                * ( real(i,lng_dbl) - 0.5_lng_dbl )
+          ref_j = -1._lng_dbl + 2._lng_dbl/real(fv_nphys,lng_dbl) & 
+                                * ( real(j,lng_dbl) - 0.5_lng_dbl )
+          !-------------------------------------------------------------------
+          ! Although this is a simpler method of defining cell centers, 
+          ! it is inconsistent with the method used by TempestRemap.
+          ! Instead use the average of cell corners in cartesian coordinates.
+          !-------------------------------------------------------------------
+          ! sphere_coord = ref2sphere(ref_i, ref_j,                         &
+          !                           elem(ie)%corners3D, cubed_sphere_map, &
+          !                           elem(ie)%corners, elem(ie)%facenum )
+          ! fv_physgrid(ie)%lat(i,j) = sphere_coord%lat
+          ! fv_physgrid(ie)%lon(i,j) = sphere_coord%lon
+          !-------------------------------------------------------------------
+          ! cell corner locations
+          !-------------------------------------------------------------------
+          center_tmp%x = 0.0_lng_dbl
+          center_tmp%y = 0.0_lng_dbl
+          center_tmp%z = 0.0_lng_dbl
+          do c = 1,4
+            ref_corner_i = ref_i + corner_offset_i(c)/real(fv_nphys,lng_dbl)
+            ref_corner_j = ref_j + corner_offset_j(c)/real(fv_nphys,lng_dbl)
+
+            ! change element local reference coordinate to spherical
+            sphere_coord = ref2sphere(ref_corner_i, ref_corner_j,           &
+                                      elem(ie)%corners3D, cubed_sphere_map, &
+                                      elem(ie)%corners, elem(ie)%facenum )
+
+            fv_physgrid(ie)%corner_lat(i,j,c) = sphere_coord%lat
+            fv_physgrid(ie)%corner_lon(i,j,c) = sphere_coord%lon
+
+            ! change spherical to cartesian
+            corner_tmp(c) = change_coordinates(sphere_coord)
+
+            ! average corner coordinates to find center
+            center_tmp%x = center_tmp%x + corner_tmp(c)%x/4._lng_dbl
```

Comment:
Does this averaging XYZ coordinates take the averaged cell center off the sphere?

Reply:
Exactly. This part of the code was actually very tricky because I needed the coordinates to match TempestRemap exactly. I originally used the cell center based on the reference coordinate, but because of round off error (I think) it never matched the domain files that were generated from the grids that came out of TempestRemap's "GenerateVolumetricMesh" routine that creates subcells. So I carefully went over the TR code and found that doing the calculation on the sphere like this solved the problem.

Reply:
Thanks. I am not sure how TempestRemap comes in here but in a long run, shouldn't we plan to have the same cell centers in physics as in tempest? That is, is it more practical to read cell centers from a file instead of recomputing them? Maybe, easiest is to make tempest read such file, because it is an offline tool.

Reply:
This might be a good idea, but this is probably a question for Mark. This might not be an issue when all components are running on different grids and mapping files become more important.